Pull Request: Trajactory and Function line visualisation

kianniek opened the pull request:
# Pull Request: Trajactory and Function line visualisation

## Related Issues

- [Feature: Trajectory Line](https://github.com/Burning-Equations/Function-Dungeon-II/issues/47)
- [Feature: Object Shake](https://github.com/Burning-Equations/Function-Dungeon-II/issues/40)
- [Feature: Cannon Platform](https://github.com/Burning-Equations/Function-Dungeon-II/issues/46)

## Summary of Changes

- Implements the Trajectory Line and function line feature described in the game design document and technical requirements analysis.

- Added Object shake behaviour script that works with any game object

- Added the cannon platform that moves vertically with limits on how hight and low it will go

## Additional Remarks

## Pull Criteria

- [x] Code commits have been reviewed
- [x] Changes have been made if requested
---
![image](https://github.com/Burning-Equations/Function-Dungeon-II/assets/56929026/20e74b5f-7598-4da1-bdf7-3cb98cf76d0c)

MisakiCopperrose reviewed (changes requested):
Currently, there are merge conflicts regarding the packages you're using and I notice you have added the following but did not mention them:

- Editor coroutines
- Profile analyzer
- SRP core
- URP
- Code coverage

I know some of these are dependent on each other and added automatically, but I believe there were problems last time with the URP package?

---

Regarding code style, I noticed the following: 

- Folders have spaces in-between and have varying naming styles instead of PascalCase
- Mixed usage of the explicit and implicit `private` access modifier
- Missing namespace declaration and usage, everything is currently in the root namespace
- Line drawer editor scripts have an `_` before editor
- Inline explicit type declaration usage over using implicit `var` (FunctionLineController_Editor, FunctionLineController, TrajectoryLineController, ShakeBehaviourEditor)
- Using standard methods instead of properties (FunctionLineController)
- LineSettings.cs has multiple classes

---

Regarding implementation:
- LineSettingsVariables is set as a serializable class, please use ScriptableObjects in this case
- In LineSettings you have duplicated behaviour by incorrectly using OnValidate(), OnValidate() should be calling ApplySettings(..), if misunderstood, please elaborate.

---

Lemme know what you think

kianniek commented:
@MisakiCopperrose 

- Packages: Fixed and merged with the project to work with URP (which wasn't installed yet)

- Regarding code style: Reworked things to fit with your findings

- Regarding implementation: Reimplemented the double class to work as a scriptable object function and workings regarding this change remain the same in the editor.

Let me know if something is not up to standard :)

MisakiCopperrose commented on Assets/Scripts/Shake/ShakeBehavior.cs:
```diff
@@ -1,123 +1,132 @@
+        [SerializeField]
+        UnityEvent OnShakeStart;
+        [SerializeField]
+        UnityEvent OnShakeEnd;
```
Inconsistency with the access modifiers and missing object creation

MisakiCopperrose commented on Assets/Scripts/LineDrawer/Editor/TrajectoryLineControllerEditor.cs:
```diff
@@ -0,0 +1,68 @@
+            serializedObject.Update();  // Load the actual values
+
+            // Script field
+            MonoScript script = MonoScript.FromMonoBehaviour((TrajectoryLineController)target);
```
Explicit type definition, implicit can be easily used here

MisakiCopperrose commented on Assets/Scripts/LineDrawer/LineSettingVarables.cs:
```diff
@@ -0,0 +1,119 @@
+    [CreateAssetMenu(fileName = "LineSettingVariables", menuName = "LineDrawer/LineSettingVariables")]
+    public class LineSettingVariables : ScriptableObject
+    {
+        public event Action OnSettingsChanged;
```
What's the reasoning for using a C# `Action` over the `UnityAction` type? 

Also, `ScriptableObject`'s have an `OnValidate()` method which is called if anything is changed from the inspector.

kianniek replied:
The Reasoning for using C# `Action` over the `UnityAction` type is because I reconed it would have an performance advantage of 0,00015 millisecond but I changed it to UnityAction now to be consistent with what action is used.

I knew OnValidate() was an option but it didn't work when I tried it but found out why OnValidate() wouldn't work for me. Fixed now so that it works with OnValidate and doesn't need an separate Editor script.

MisakiCopperrose reviewed (changes requested):
Ples don't hit me tomorrow XD

I found out how to put in-line comments to better guide where the problems are, so you don't have to go an search for them :D

Think you just accidentally looked over 'em
